File: android/testSrc/org/jetbrains/android/refactoring/AndroidFindStyleApplicationsTest.java
package org.jetbrains.android.refactoring;

import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.psi.xml.XmlFile;
import com.intellij.psi.xml.XmlTag;
import com.intellij.usageView.UsageInfo;
import com.intellij.util.xml.DomManager;
import org.jetbrains.android.AndroidTestCase;
import org.jetbrains.android.dom.layout.LayoutDomFileDescription;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * @author Eugene.Kudelevsky
 */
public class AndroidFindStyleApplicationsTest extends AndroidTestCase {
  private static final String BASE_PATH = "refactoring/findPossibleUsages/";

  public void test1() throws Exception {
    doTest();
  }

  public void testGranular1() throws Exception {
    myFixture.copyFileToProject(BASE_PATH + "1_layout.xml", "res/layout/layout.xml");
    final VirtualFile f = myFixture.copyFileToProject(BASE_PATH + "1.xml", "res/values/styles.xml");
    myFixture.configureFromExistingVirtualFile(f);
    XmlTag tag = PsiTreeUtil.getParentOfType(myFixture.getElementAtCaret(), XmlTag.class);
    AndroidFindStyleApplicationsAction.MyStyleData styleData = AndroidFindStyleApplicationsAction.getStyleData(tag);
    assertNotNull(styleData);
    AndroidFindStyleApplicationsProcessor processor =
      AndroidFindStyleApplicationsAction.createFindStyleApplicationsProcessor(tag, styleData, null);
    processor.configureScope(AndroidFindStyleApplicationsProcessor.MyScope.PROJECT, null);
    Collection<PsiFile> files = processor.collectFilesToProcess();
    assertEquals(1, files.size());
    XmlFile layoutFile = (XmlFile)files.iterator().next();
    assertInstanceOf(DomManager.getDomManager(myFixture.getProject()).getDomFileDescription(
      (XmlFile)layoutFile), LayoutDomFileDescription.class);
    final List<UsageInfo> usages = new ArrayList<>();
    processor.collectPossibleStyleApplications(layoutFile, usages);
    assertEquals(2, usages.size());
  }

  public void test2() throws Exception {
    doTest();
  }

  public void test3() throws Exception {
    doTest();
  }

  public void test4() throws Exception {
    doTest();
  }

  public void test5() throws Exception {
    doTest();
  }

  public void test6() throws Exception {
    final String testName = getTestName(true);
    myFixture.copyFileToProject(BASE_PATH + testName + "_layout.xml", "res/layout/layout1.xml");
    myFixture.copyFileToProject(BASE_PATH + testName + "_layout.xml", "res/layout/layout2.xml");

    doTest1();

    myFixture.checkResultByFile("res/layout/layout1.xml", BASE_PATH + testName + "_layout_after.xml", true);
    myFixture.checkResultByFile("res/layout/layout2.xml", BASE_PATH + testName + "_layout_after.xml", true);
  }

  public void test7() throws Exception {
    try {
      doTest();
      fail();
    }
    catch (RuntimeException e) {
      assertEquals("IDEA has not found any possible applications of style 'style1'", e.getMessage());
    }
  }

  private void doTest() {
    final String testName = getTestName(true);
    myFixture.copyFileToProject(BASE_PATH + testName + "_layout.xml", "res/layout/layout.xml");

    doTest1();

    myFixture.checkResultByFile("res/layout/layout.xml", BASE_PATH + testName + "_layout_after.xml", true);
  }

  private void doTest1() {
    final String testName = getTestName(true);
    final VirtualFile f = myFixture.copyFileToProject(BASE_PATH + testName + ".xml", "res/values/styles.xml");
    myFixture.configureFromExistingVirtualFile(f);
    myFixture.testAction(new AndroidFindStyleApplicationsAction(new AndroidFindStyleApplicationsAction.MyTestConfig(
      AndroidFindStyleApplicationsProcessor.MyScope.PROJECT)));
    myFixture.checkResultByFile(BASE_PATH + testName + ".xml");
  }
}

File: editing/testSrc/org/jetbrains/android/refactoring/AndroidFindStyleApplicationsTest.kt
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jetbrains.android.refactoring

import com.android.testutils.TestUtils
import com.android.tools.idea.testing.AndroidProjectRule
import com.google.common.truth.Truth.assertThat
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.runReadAction
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.xml.XmlFile
import com.intellij.psi.xml.XmlTag
import com.intellij.usageView.UsageInfo
import com.intellij.util.xml.DomManager
import org.jetbrains.android.dom.layout.LayoutViewElementDomFileDescription
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import kotlin.test.assertFailsWith

private const val BASE_PATH = "refactoring/findStyleApplication/"

@RunWith(JUnit4::class)
class AndroidFindStyleApplicationsTest {
  @get:Rule
  var androidProjectRule: AndroidProjectRule = AndroidProjectRule.onDisk()

  private val myFixture by lazy { androidProjectRule.fixture }

  @Before
  fun setUp() {
    myFixture.testDataPath = TestUtils.resolveWorkspacePath("tools/adt/idea/editing/testData").toString()
  }

  @Test
  fun test1() {
    doTest("1")
  }

  @Test
  fun testGranular1() {
    myFixture.copyFileToProject(BASE_PATH + "1_layout.xml", "res/layout/layout.xml")
    val styleVirtualFile = myFixture.copyFileToProject(BASE_PATH + "1.xml", "res/values/styles.xml")
    myFixture.configureFromExistingVirtualFile(styleVirtualFile)

    val tag = runReadAction { PsiTreeUtil.getParentOfType(myFixture.file.findElementAt(myFixture.caretOffset), XmlTag::class.java) }
    assertThat(tag).isNotNull()

    val styleData = runReadAction { AndroidFindStyleApplicationsAction.getStyleData(tag!!) }
    assertThat(styleData).isNotNull()

    val processor = runReadAction { AndroidFindStyleApplicationsAction.createFindStyleApplicationsProcessor(tag, styleData, null) }
    processor.configureScope(AndroidFindStyleApplicationsProcessor.MyScope.PROJECT, null)

    val files = runReadAction { processor.collectFilesToProcess() }
    assertThat(files).hasSize(1)

    val layoutFile = files.single() as XmlFile
    val domFileDescription = runReadAction { DomManager.getDomManager(myFixture.project).getDomFileDescription(layoutFile) }
    assertThat(domFileDescription)
      .isInstanceOf(LayoutViewElementDomFileDescription::class.java)

    val usages: List<UsageInfo> = ArrayList()
    runReadAction { processor.collectPossibleStyleApplications(layoutFile, usages) }
    assertThat(usages).hasSize(2)
  }

  @Test
  fun test2() {
    doTest("2")
  }

  @Test
  fun test3() {
    doTest("3")
  }

  @Test
  fun test4() {
    doTest("4")
  }

  @Test
  fun test5() {
    doTest("5")
  }

  @Test
  fun test6() {
    myFixture.copyFileToProject(BASE_PATH + "6_layout.xml", "res/layout/layout1.xml")
    myFixture.copyFileToProject(BASE_PATH + "6_layout.xml", "res/layout/layout2.xml")
    doTest1("6")
    myFixture.checkResultByFile("res/layout/layout1.xml", BASE_PATH + "6_layout_after.xml", true)
    myFixture.checkResultByFile("res/layout/layout2.xml", BASE_PATH + "6_layout_after.xml", true)
  }

  @Test
  fun test7() {
    assertFailsWith<RuntimeException>("IDEA has not found any possible applications of style 'style1'") { doTest("7") }
  }

  private fun doTest(testName: String) {
    myFixture.copyFileToProject("$BASE_PATH${testName}_layout.xml", "res/layout/layout.xml")
    doTest1(testName)
    myFixture.checkResultByFile("res/layout/layout.xml", "$BASE_PATH${testName}_layout_after.xml", true)
  }

  private fun doTest1(testName: String) {
    val f = myFixture.copyFileToProject("$BASE_PATH$testName.xml", "res/values/styles.xml")
    myFixture.configureFromExistingVirtualFile(f)

    val myTestConfig = AndroidFindStyleApplicationsAction.MyTestConfig(AndroidFindStyleApplicationsProcessor.MyScope.PROJECT)
    val action = AndroidFindStyleApplicationsAction(myTestConfig)
    ApplicationManager.getApplication().invokeAndWait { myFixture.testAction (action) }

    myFixture.checkResultByFile("$BASE_PATH$testName.xml")
  }
}
